File: src/bin/18_strong_components.py
from collections import defaultdict, deque
from itertools import groupby
from pprint import pprint
from tokenize import group


Graph = dict[str, set[str]]


def rev(graph: Graph) -> Graph:
  rev_graph: Graph = defaultdict(set)

  for a, edges in graph.items():
    for b in edges:
      rev_graph[b].add(a)

  return rev_graph


def strong_components(graph: Graph):
  rev_graph = rev(graph)
  visited = set[str]()
  queue = deque[str]()
  node_label = dict[str, int]()

  def postorder_dfs(node: str):
    visited.add(node)

    for neighbour in rev_graph[node]:  # REVERSE
      if neighbour not in visited:
        postorder_dfs(neighbour)

    queue.appendleft(node)

  for node in graph:
    if node not in visited:
      postorder_dfs(node)

  def label_dfs(node: str, label: int):
    node_label[node] = label

    for neighbour in graph[node]:  # FORWARD
      if neighbour not in node_label:
        label_dfs(neighbour, label)

  label = 0
  for node in queue:
    if node not in node_label:
      label_dfs(node, label)
      label += 1

  components = sorted((label, node) for node, label in node_label.items())
  components = groupby(components, key=lambda t: t[0])
  return {label: [n for l, n in group] for label, group in components}


graph: Graph = {
    "a": {"b"},
    "b": {"f"},
    "c": {"h"},
    "d": {"c"},
    "e": {"f", "i"},
    "f": {"g", "l"},
    "g": {"c", "a"},
    "h": {"l", "d"},
    "i": {"n"},
    "j": {"k", "m"},
    "k": {"h", "l"},
    "l": {"o", "p"},
    "m": {"i"},
    "n": {"j", "o"},
    "o": {"k"},
    "p": set(),
}

pprint(strong_components(graph))

File: src/bin/18_strong_components.rs
use std::collections::{BTreeMap, HashMap, HashSet, VecDeque};

type Graph = HashMap<char, HashSet<char>>;

fn postorder_dfs(
    graph: &Graph,
    deque: &mut VecDeque<char>,
    visited: &mut HashSet<char>,
    node: &char,
) {
    visited.insert(*node);
    for neighbour in graph.get(&node).unwrap_or(&HashSet::new()).iter() {
        if !visited.contains(neighbour) {
            postorder_dfs(graph, deque, visited, neighbour);
        }
    }
    deque.push_front(*node);
}

fn label_dfs(graph: &Graph, node_label: &mut HashMap<char, usize>, node: &char, label: usize) {
    node_label.insert(*node, label);

    for neighbour in graph[node].iter() {
        if !node_label.contains_key(neighbour) {
            label_dfs(graph, node_label, neighbour, label);
        }
    }
}

fn strong_components(graph: &Graph) -> BTreeMap<usize, HashSet<char>> {
    let rev_graph = rev(graph);
    let mut visited = HashSet::<char>::new();
    let mut deque = VecDeque::<char>::new();
    let mut node_label = HashMap::<char, usize>::new();

    for node in graph.keys() {
        if !visited.contains(node) {
            postorder_dfs(&rev_graph, &mut deque, &mut visited, node);
        }
    }

    let mut label: usize = 0;
    for node in deque.iter() {
        if !node_label.contains_key(node) {
            label_dfs(graph, &mut node_label, node, label);
            label += 1;
        }
    }

    let mut components: BTreeMap<usize, HashSet<char>> = BTreeMap::new();
    for (node, label) in node_label {
        components.entry(label).or_default().insert(node);
    }
    return components;
}

fn rev(graph: &Graph) -> Graph {
    let mut rev_graph = HashMap::<char, HashSet<char>>::new();

    for (a, edges) in graph.iter() {
        for b in edges.iter() {
            rev_graph.entry(*b).or_default().insert(*a);
        }
    }

    return rev_graph;
}

fn main() {
    let graph = Graph::from([
        ('a', HashSet::from(['b'])),
        ('b', HashSet::from(['f'])),
        ('c', HashSet::from(['h'])),
        ('d', HashSet::from(['c'])),
        ('e', HashSet::from(['f', 'i'])),
        ('f', HashSet::from(['g', 'l'])),
        ('g', HashSet::from(['c', 'a'])),
        ('h', HashSet::from(['l', 'd'])),
        ('i', HashSet::from(['n'])),
        ('j', HashSet::from(['k', 'm'])),
        ('k', HashSet::from(['h', 'l'])),
        ('l', HashSet::from(['o', 'p'])),
        ('m', HashSet::from(['i'])),
        ('n', HashSet::from(['j', 'o'])),
        ('o', HashSet::from(['k'])),
        ('p', HashSet::from([])),
    ]);

    let components = strong_components(&graph);

    println!("{components:#?}");
}
